File: agent/projects/store.py
from __future__ import annotations

import hashlib
import json
import re
from datetime import datetime, timedelta, timezone
from pathlib import Path
from typing import Any

_UTC8 = timezone(timedelta(hours=8))
_VERSION = 1

PROJECT_MEMORY_START = "<!-- emperor-agent:project-memory:start -->"
PROJECT_MEMORY_END = "<!-- emperor-agent:project-memory:end -->"
_DEFAULT_BLOCK = "## Emperor Agent Project Memory\n\n- 尚未记录项目情况。"


class ProjectStore:
    """Local project registry and AGENTS.md managed memory block."""

    def __init__(self, root: Path) -> None:
        self.root = Path(root).resolve()
        self.projects_dir = self.root / "memory" / "projects"
        self.index_path = self.projects_dir / "index.json"

    def resolve(self, path: str | Path) -> dict[str, Any]:
        project_path = Path(path).expanduser().resolve()
        if not project_path.exists() or not project_path.is_dir():
            raise ValueError("project path must be an existing directory")
        self._ensure_agents(project_path)
        entry = self._entry_for_path(project_path)
        loaded = self._load()
        items = [item for item in loaded if item.get("project_id") != entry["project_id"]]
        existing = next((item for item in loaded if item.get("project_id") == entry["project_id"]), {})
        entry["summary"] = str(existing.get("summary") or "")
        entry["created_at"] = str(existing.get("created_at") or entry["created_at"])
        items.append(entry)
        self._save_sorted(items)
        return dict(entry)

    def get(self, project_id: str) -> dict[str, Any] | None:
        for item in self._load():
            if item.get("project_id") == project_id:
                return dict(item)
        return None

    def list(self) -> list[dict[str, Any]]:
        return sorted(self._load(), key=lambda item: str(item.get("updated_at") or ""), reverse=True)

    def read_agents(self, project_id: str) -> str:
        entry = self.get(project_id)
        if not entry:
            return ""
        path = Path(str(entry.get("project_path") or "")) / "AGENTS.md"
        if not path.exists():
            return ""
        return path.read_text(encoding="utf-8")

    def read_managed_memory(self, project_id: str) -> str:
        return _extract_block(self.read_agents(project_id)) or ""

    def update_memory(self, project_id: str, content: str) -> dict[str, Any]:
        entry = self.get(project_id)
        if entry is None:
            raise KeyError(f"unknown project: {project_id}")
        project_path = Path(str(entry["project_path"]))
        agents_path = project_path / "AGENTS.md"
        self._ensure_agents(project_path)
        current = agents_path.read_text(encoding="utf-8")
        text = _replace_block(current, content.strip() or _DEFAULT_BLOCK)
        agents_path.write_text(text.rstrip() + "\n", encoding="utf-8")

        summary = _summarize(content)
        updated = dict(entry)
        updated["summary"] = summary
        updated["updated_at"] = self._stamp()
        items = [item for item in self._load() if item.get("project_id") != project_id]
        items.append(updated)
        self._save_sorted(items)
        return updated

    def summary_for_chat(self, *, limit: int = 8) -> str:
        lines: list[str] = []
        for item in self.list()[:limit]:
            name = str(item.get("project_name") or item.get("project_path") or "project")
            summary = str(item.get("summary") or "").strip()
            path = str(item.get("project_path") or "")
            label = f"{name} ({path})" if path else name
            lines.append(f"- {label}: {summary or '已绑定为 Build 项目'}")
        return "\n".join(lines)

    def _entry_for_path(self, project_path: Path) -> dict[str, Any]:
        project_id = hashlib.sha256(str(project_path).encode("utf-8")).hexdigest()[:16]
        now = self._stamp()
        return {
            "project_id": project_id,
            "project_path": str(project_path),
            "project_name": project_path.name or str(project_path),
            "summary": "",
            "created_at": now,
            "updated_at": now,
            "agents_path": str(project_path / "AGENTS.md"),
            "version": _VERSION,
        }

    def _ensure_agents(self, project_path: Path) -> None:
        agents_path = project_path / "AGENTS.md"
        if not agents_path.exists():
            agents_path.write_text(
                "# AGENTS.md\n\n"
                "本文件记录该项目给 Agent 的协作规则和项目记忆。\n\n"
                f"{PROJECT_MEMORY_START}\n{_DEFAULT_BLOCK}\n{PROJECT_MEMORY_END}\n",
                encoding="utf-8",
            )
            return
        text = agents_path.read_text(encoding="utf-8")
        if PROJECT_MEMORY_START in text and PROJECT_MEMORY_END in text:
            return
        agents_path.write_text(
            text.rstrip()
            + "\n\n"
            + f"{PROJECT_MEMORY_START}\n{_DEFAULT_BLOCK}\n{PROJECT_MEMORY_END}\n",
            encoding="utf-8",
        )

    def _load(self) -> list[dict[str, Any]]:
        if not self.index_path.exists():
            return []
        try:
            data = json.loads(self.index_path.read_text(encoding="utf-8") or "[]")
        except (json.JSONDecodeError, OSError):
            return []
        if not isinstance(data, list):
            return []
        return [item for item in data if isinstance(item, dict)]

    def _save_sorted(self, items: list[dict[str, Any]]) -> None:
        items.sort(key=lambda item: str(item.get("updated_at") or ""), reverse=True)
        self.projects_dir.mkdir(parents=True, exist_ok=True)
        tmp = self.index_path.with_suffix(".json.tmp")
        tmp.write_text(json.dumps(items, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")
        tmp.replace(self.index_path)

    def _stamp(self) -> str:
        return datetime.now(_UTC8).strftime("%Y-%m-%dT%H:%M:%S%z")


def _extract_block(text: str) -> str | None:
    start = text.find(PROJECT_MEMORY_START)
    end = text.find(PROJECT_MEMORY_END)
    if start < 0 or end < 0 or end < start:
        return None
    body_start = start + len(PROJECT_MEMORY_START)
    return text[body_start:end].strip()


def _replace_block(text: str, content: str) -> str:
    start = text.find(PROJECT_MEMORY_START)
    end = text.find(PROJECT_MEMORY_END)
    if start < 0 or end < 0 or end < start:
        return text.rstrip() + "\n\n" + f"{PROJECT_MEMORY_START}\n{content}\n{PROJECT_MEMORY_END}"
    body_start = start + len(PROJECT_MEMORY_START)
    return text[:body_start].rstrip() + "\n" + content.strip() + "\n" + text[end:].lstrip()


def _summarize(content: str) -> str:
    lines = []
    for line in str(content or "").splitlines():
        stripped = line.strip()
        if not stripped or stripped.startswith("#"):
            continue
        stripped = re.sub(r"^[-*+\d.)\s]+", "", stripped).strip()
        if stripped:
            lines.append(stripped)
    parts = [part.strip(" \t\r\n。；;") for part in re.split(r"[\n。；;]+", "\n".join(lines)) if part.strip(" \t\r\n。；;")]
    return "；".join(parts)[:120]

File: packages/core/src/projects/store.ts
import { createHash } from 'node:crypto'
import {
  existsSync,
  mkdirSync,
  readFileSync,
  renameSync,
  statSync,
  writeFileSync,
} from 'node:fs'
import { basename, join, resolve } from 'node:path'
import {
  DEFAULT_PROJECT_MEMORY_BLOCK,
  ProjectStateStore,
  type ProjectStateMetadata,
} from './state-store'
import {
  appendPatchLedger,
  applyMemoryPatch,
  memoryContentHash,
  type MemoryPatchOperation,
} from '../memory/patch'
import type { MemoryVersionStore } from '../memory/versions'

export { PROJECT_MEMORY_END, PROJECT_MEMORY_START } from './state-store'

const VERSION = 1

export interface LegacyProjectPrivateData {
  sessions: boolean
  memory: boolean
}

export interface ProjectEntry extends ProjectStateMetadata {
  project_id: string
  project_path: string
  workspace_path: string
  project_name: string
  summary: string
  created_at: string
  updated_at: string
  agents_path: string
  state_path: string
  memory_path: string
  project_json_path: string
  prompt_overlay_path: string
  legacy_agents_path: string | null
  legacy_imported_at: string | null
  version: number
}

export class ProjectStore {
  readonly root: string
  readonly projectsDir: string
  readonly indexPath: string
  readonly stateStore: ProjectStateStore
  private readonly versions: Pick<
    MemoryVersionStore,
    'snapshotPath' | 'nextVersionForPath'
  > | null

  constructor(
    root: string,
    opts: {
      versions?: Pick<
        MemoryVersionStore,
        'snapshotPath' | 'nextVersionForPath'
      > | null
    } = {},
  ) {
    this.root = resolve(root)
    this.projectsDir = join(this.root, 'projects')
    this.indexPath = join(this.projectsDir, 'index.json')
    this.stateStore = new ProjectStateStore(this.projectsDir)
    this.versions = opts.versions ?? null
  }

  resolve(path: string): ProjectEntry {
    const projectPath = resolve(path)
    if (!existsSync(projectPath) || !statSync(projectPath).isDirectory()) {
      throw new Error('project path must be an existing directory')
    }
    const entry = this.entryForPath(projectPath)
    const loaded = this.load()
    const existing = loaded.find((item) => item.project_id === entry.project_id)
    const next: ProjectEntry = {
      ...entry,
      summary: existing?.summary ?? '',
      created_at: existing?.created_at ?? entry.created_at,
    }
    const ensured = this.stateStore.ensureProject(next)
    this.saveSorted([
      ...loaded.filter((item) => item.project_id !== entry.project_id),
      ensured,
    ])
    return { ...ensured }
  }

  get(projectId: string): ProjectEntry | null {
    return this.load().find((item) => item.project_id === projectId) ?? null
  }

  /** Detects private `.emperor/sessions` or `.emperor/memory` already sitting inside the
   * project's own source tree (from an older layout or another tool). Diagnostics-only:
   * this never deletes or migrates that data automatically — see global-state-store plan
   * Task 8. */
  detectLegacyPrivateData(projectPath: string): LegacyProjectPrivateData {
    const dotEmperor = join(resolve(projectPath), '.emperor')
    return {
      sessions: existsSync(join(dotEmperor, 'sessions')),
      memory: existsSync(join(dotEmperor, 'memory')),
    }
  }

  list(): ProjectEntry[] {
    return this.load().sort((a, b) => b.updated_at.localeCompare(a.updated_at))
  }

  readAgents(projectId: string): string {
    const entry = this.get(projectId)
    if (!entry) return ''
    return [
      this.stateStore.readAgents(entry.project_id),
      this.stateStore.readWorkspaceCollaborationContext(entry.project_id),
    ]
      .filter((part) => part.trim())
      .join('\n\n---\n\n')
  }

  readManagedMemory(projectId: string): string {
    return this.stateStore.readManagedMemory(projectId)
  }

  updateMemory(projectId: string, content: string): ProjectEntry {
    const entry = this.get(projectId)
    if (!entry) throw new Error(`unknown project: ${projectId}`)
    this.stateStore.ensureProject(entry)
    const current = normalizeProjectMemoryBase(
      this.stateStore.readManagedMemory(projectId),
    )
    const operations = projectMemorySectionReplacementOps(content)
    if (!operations.length)
      throw new Error('project memory update requires at least one ## section')
    const version = this.versions
      ? this.versions.nextVersionForPath(entry.agents_path, {
          target: 'project',
        })
      : 1
    const patch = {
      target: { kind: 'project' as const, projectId },
      baseVersion: version,
      baseHash: memoryContentHash(current),
      operations,
      rationale: 'save_project_memory',
    }
    const result = applyMemoryPatch(patch, current, {
      mode: 'build',
      explicitReplace: true,
      currentVersion: this.versions ? version : undefined,
    })
    if (!result.ok)
      throw new Error(
        `project memory update rejected: ${result.errors.join(', ')}`,
      )
    if (this.versions && existsSync(entry.agents_path)) {
      this.versions.snapshotPath(entry.agents_path, {
        target: 'project',
        reason: 'save_project_memory',
      })
    }
    this.stateStore.writeManagedMemory(projectId, result.content)
    appendPatchLedger(
      join(this.root, 'memory', 'patch-ledger.jsonl'),
      patch,
      current,
      result,
    )

    const updated: ProjectEntry = {
      ...entry,
      summary: summarize(result.content),
      updated_at: stamp(),
    }
    this.stateStore.writeProjectJson(updated)
    this.saveSorted([
      ...this.load().filter((item) => item.project_id !== projectId),
      updated,
    ])
    return { ...updated }
  }

  summaryForChat(opts: { limit?: number } = {}): string {
    const limit = opts.limit ?? 8
    const lines: string[] = []
    for (const item of this.list().slice(0, limit)) {
      const name = item.project_name || item.project_path || 'project'
      const label = item.project_path ? `${name} (${item.project_path})` : name
      lines.push(`- ${label}: 已绑定为 Build 项目`)
    }
    return lines.join('\n')
  }

  private entryForPath(projectPath: string): ProjectEntry {
    const projectId = createHash('sha256')
      .update(projectPath, 'utf8')
      .digest('hex')
      .slice(0, 16)
    const now = stamp()
    const paths = this.stateStore.paths(projectId)
    return {
      project_id: projectId,
      project_path: projectPath,
      workspace_path: projectPath,
      project_name: basename(projectPath) || projectPath,
      summary: '',
      created_at: now,
      updated_at: now,
      ...paths,
      legacy_agents_path: null,
      legacy_imported_at: null,
      version: VERSION,
    }
  }

  private load(): ProjectEntry[] {
    if (!existsSync(this.indexPath)) return []
    try {
      const data = JSON.parse(readFileSync(this.indexPath, 'utf8') || '[]')
      return Array.isArray(data)
        ? data
            .filter(isObject)
            .map((item) => normalizeProject(item, this.projectsDir))
        : []
    } catch {
      return []
    }
  }

  private saveSorted(items: ProjectEntry[]): void {
    items.sort((a, b) => b.updated_at.localeCompare(a.updated_at))
    mkdirSync(this.projectsDir, { recursive: true })
    const tmp = this.indexPath.replace(/\.json$/, '.json.tmp')
    writeFileSync(tmp, JSON.stringify(items, null, 2) + '\n', 'utf8')
    renameSync(tmp, this.indexPath)
  }
}

function summarize(content: string): string {
  const lines: string[] = []
  for (const line of String(content || '').split('\n')) {
    let stripped = line.trim()
    if (!stripped || stripped.startsWith('#')) continue
    stripped = stripped.replace(/^[-*+\d.)\s]+/, '').trim()
    if (stripped) lines.push(stripped)
  }
  return lines
    .join('\n')
    .split(/[\n。；;]+/)
    .map((part) => part.trim().replace(/[ \t\r\n。；;]+$/g, ''))
    .filter(Boolean)
    .join('；')
    .slice(0, 120)
}

function normalizeProjectMemoryBase(content: string): string {
  return String(content ?? '').trim() === DEFAULT_PROJECT_MEMORY_BLOCK.trim()
    ? ''
    : content
}

function projectMemorySectionReplacementOps(
  markdown: string,
): MemoryPatchOperation[] {
  const text = `${String(markdown || '').trimEnd()}\n`
  const matches = [...text.matchAll(/^##\s+(.+?)\s*$/gm)]
  return matches.map((match, index) => {
    const section = String(match[1] ?? '').trim()
    const bodyStart = (match.index ?? 0) + match[0].length
    const bodyEnd =
      index + 1 < matches.length
        ? (matches[index + 1]!.index ?? text.length)
        : text.length
    return {
      op: 'replace_section',
      section,
      content: text.slice(bodyStart, bodyEnd).trim(),
    }
  })
}

function normalizeProject(
  raw: Record<string, unknown>,
  projectsDir: string,
): ProjectEntry {
  const projectId = String(raw.project_id ?? '')
  const stateStore = new ProjectStateStore(projectsDir)
  const paths = stateStore.paths(projectId)
  return {
    project_id: projectId,
    project_path: String(raw.project_path ?? ''),
    workspace_path: String(raw.workspace_path ?? raw.project_path ?? ''),
    project_name: String(raw.project_name ?? ''),
    summary: String(raw.summary ?? ''),
    created_at: String(raw.created_at ?? ''),
    updated_at: String(raw.updated_at ?? ''),
    ...paths,
    legacy_agents_path: nullableText(raw.legacy_agents_path),
    legacy_imported_at: nullableText(raw.legacy_imported_at),
    version: Number(raw.version ?? VERSION) || VERSION,
  }
}

function isObject(value: unknown): value is Record<string, unknown> {
  return Boolean(value && typeof value === 'object' && !Array.isArray(value))
}

function nullableText(value: unknown): string | null {
  return typeof value === 'string' && value.trim() ? value.trim() : null
}

function stamp(): string {
  const d = new Date(Date.now() + 8 * 3600 * 1000)
  const pad = (n: number) => String(n).padStart(2, '0')
  return `${d.getUTCFullYear()}-${pad(d.getUTCMonth() + 1)}-${pad(d.getUTCDate())}T${pad(d.getUTCHours())}:${pad(d.getUTCMinutes())}:${pad(d.getUTCSeconds())}+0800`
}
